VMware: Add support in VMWare modules for BIOS and instance UUID's

naphta opened the pull request:
##### SUMMARY
Adding support for specifying the type of UUID to match against, by default it will continue to use the bios UUID which originally was the only one which worked.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, task or feature -->
lib/ansible/module_utils/vmware.py
lib/ansible/modules/cloud/vmware/vmware_guest_boot_facts.py
lib/ansible/modules/cloud/vmware/vmware_guest_boot_manager.py
lib/ansible/modules/cloud/vmware/vmware_guest_custom_attributes.py
lib/ansible/modules/cloud/vmware/vmware_guest_disk_facts.py
lib/ansible/modules/cloud/vmware/vmware_guest_facts.py
lib/ansible/modules/cloud/vmware/vmware_guest_file_operation.py
lib/ansible/modules/cloud/vmware/vmware_guest_find.py
lib/ansible/modules/cloud/vmware/vmware_guest_move.py
lib/ansible/modules/cloud/vmware/vmware_guest_powerstate.py
lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py
lib/ansible/modules/cloud/vmware/vmware_guest_snapshot_facts.py
lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py
lib/ansible/modules/cloud/vmware/vmware_vm_shell.py
lib/ansible/modules/cloud/vmware/vmware_vmotion.py
test/integration/targets/vmware_guest_facts/tasks/main.yml

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.7.0.dev0
  config file = None
  configured module search path = [u'/Users/jhill/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.15 (default, Jun 17 2018, 12:46:58) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)]
```


##### ADDITIONAL INFORMATION
I've put this in (and to every referenced module) to increase the options of how you can refer to a VM. In-house where I work we created a lot of VMWare modules before they were available directly from Ansible so in doing this we personally can migrate to Ansible code more easily. Broadly speaking it seemed logical to support either kind of UUID as in PyVmomi it boils down to a true/false to determine which one it searched on.

naphta committed:
Add support for changing UUID type referenced

naphta committed:
Update all appropriate VMware modules to include UUID type

naphta committed:
Add integration test for filtering on instance UUID

naphta committed:
Fix issues raised from shippable test

naphta committed:
Fix incorrect variable on assertion

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_boot_manager.py:
```diff
@@ -38,6 +38,12 @@
      description:
      - UUID of the instance to manage if known, this is VMware's BIOS UUID.
```
`... BIOS UUID by default.`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_boot_manager.py:
```diff
@@ -38,6 +38,12 @@
      description:
      - UUID of the instance to manage if known, this is VMware's BIOS UUID.
      - This is required if C(name) parameter is not supplied.
+   uuid_type:
```
Missing `self.uuid_type = self.params['uuid_type']` in `__init__`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_boot_manager.py:
```diff
@@ -149,7 +155,8 @@ def _get_vm(self):
         if self.uuid:
-            vm_obj = find_vm_by_id(self.content, vm_id=self.uuid, vm_id_type="uuid")
+            vm_obj = find_vm_by_id(self.content, vm_id=self.uuid, vm_id_type="uuid",
+                                   vm_uuid_type="uuid")
```
`vm_uuid_type=self.uuid_type`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_custom_attributes.py:
```diff
@@ -48,6 +48,12 @@
      description:
      - UUID of the virtual machine to manage if known. This is VMware's unique identifier.
      - This is required parameter, if C(name) is not supplied.
+   uuid_type:
```
You are not doing anything with the type?
 Needs at least `self.uuid_type = self.params['uuid_type']` in `__init__`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_custom_attributes.py:
```diff
@@ -48,6 +48,12 @@
      description:
      - UUID of the virtual machine to manage if known. This is VMware's unique identifier.
```
`This is VMware's BIOS UUID by default.`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_disk_facts.py:
```diff
@@ -38,6 +38,12 @@
      description:
      - UUID of the instance to gather facts if known, this is VMware's unique identifier.
      - This is required parameter, if parameter C(name) is not supplied.
+   uuid_type:
```
You are not doing anything with the type?
 Needs at least `self.uuid_type = self.params['uuid_type']` in `__init__`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_disk_facts.py:
```diff
@@ -38,6 +38,12 @@
      description:
      - UUID of the instance to gather facts if known, this is VMware's unique identifier.
```
`this is VMware's BIOS UUID by default.`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_facts.py:
```diff
@@ -41,6 +41,12 @@
         description:
             - UUID of the instance to manage if known, this is VMware's unique identifier.
             - This is required if name is not supplied.
+   uuid_type:
```
You are not doing anything with the type?
 Needs at least `self.uuid_type = self.params['uuid_type']` in `__init__`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_facts.py:
```diff
@@ -41,6 +41,12 @@
         description:
             - UUID of the instance to manage if known, this is VMware's unique identifier.
```
`this is VMware's BIOS UUID by default.`

pdellaert commented on lib/ansible/module_utils/vmware.py:
```diff
@@ -857,7 +861,8 @@ def get_vm(self):
 
         if self.params['uuid']:
-            vm_obj = find_vm_by_id(self.content, vm_id=self.params['uuid'], vm_id_type="uuid")
+            vm_obj = find_vm_by_id(self.content, vm_id=self.params['uuid'], vm_id_type="uuid",
```
You need to check whether `self.params['uuid_type']`actually exists, if a module forgets to implement it, you'll get exceptions

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_find.py:
```diff
@@ -36,6 +36,12 @@
      description:
      - UUID of the instance to manage if known, this is VMware's BIOS UUID.
```
`... this is VMware's BIOS UUID by default.`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_move.py:
```diff
@@ -37,6 +37,12 @@
         description:
             - UUID of the virtual machine to manage if known, this is VMware's unique identifier.
             - This is required if C(name) is not supplied.
+   uuid_type:
```
Missing `self.uuid_type = self.params['uuid_type']` in `__init__`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_move.py:
```diff
@@ -37,6 +37,12 @@
         description:
             - UUID of the virtual machine to manage if known, this is VMware's unique identifier.
```
`... this is VMware's BIOS UUID by default.`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_file_operation.py:
```diff
@@ -68,6 +68,12 @@
             - 'dns_name'
             - 'inventory_path'
             - 'vm_name'
+    vm_uuid_type:
```
Doesn't it make much more sense to re-use the vm_id_type field here and just add `instance_uuid` (`uuid`) is already present. This way you can reuse that field (just make sure you update the code appropriately)

Otherwise: 
Missing `self.uuid_type = self.params['uuid_type']` in `__init__`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_powerstate.py:
```diff
@@ -42,6 +42,12 @@
     description:
     - UUID of the instance to manage if known, this is VMware's unique identifier.
```
`... this is VMware's BIOS UUID by default.`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_powerstate.py:
```diff
@@ -42,6 +42,12 @@
     description:
     - UUID of the instance to manage if known, this is VMware's unique identifier.
     - This is required if name is not supplied.
+  uuid_type:
```
Missing `self.uuid_type = self.params['uuid_type']` in `__init__`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py:
```diff
@@ -54,6 +54,12 @@
      description:
      - UUID of the instance to manage if known, this is VMware's unique identifier.
```
`... this is VMware's BIOS UUID by default.`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py:
```diff
@@ -54,6 +54,12 @@
      description:
      - UUID of the instance to manage if known, this is VMware's unique identifier.
      - This is required parameter, if C(name) is not supplied.
+   uuid_type:
```
Missing `self.uuid_type = self.params['uuid_type']` in `__init__`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_snapshot_facts.py:
```diff
@@ -38,6 +38,12 @@
      - UUID of the instance to manage if known, this value is VMware's unique identifier.
```
`... this value is VMware's BIOS UUID by default.`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_snapshot_facts.py:
```diff
@@ -38,6 +38,12 @@
      - UUID of the instance to manage if known, this value is VMware's unique identifier.
      - This is required if C(name) is not supplied.
      - The C(folder) is ignored, if C(uuid) is provided.
+   uuid_type:
```
Missing `self.uuid_type = self.params['uuid_type']` in `__init__`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py:
```diff
@@ -56,6 +56,12 @@
      description:
      - UUID of the VM  for which to wait until the tools become available, if known. This is VMware's unique identifier.
```
`this is VMware's BIOS UUID by default.`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py:
```diff
@@ -56,6 +56,12 @@
      description:
      - UUID of the VM  for which to wait until the tools become available, if known. This is VMware's unique identifier.
      - This is required, if C(name) is not supplied.
+   uuid_type:
```
Missing `self.uuid_type = self.params['uuid_type']` in `__init__`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_vm_shell.py:
```diff
@@ -63,6 +63,12 @@
       - The VMware identification method by which the virtual machine will be identified.
       default: vm_name
       choices: ['uuid', 'dns_name', 'inventory_path', 'vm_name']
+    vm_uuid_type:
```
Makes more sense to use the vm_id_type field, and just add the option `instance_uuid`. Also adjust code, of course

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_vmotion.py:
```diff
@@ -47,6 +47,12 @@
       - This is a required parameter, if C(vm_name) is not set.
```
Line above: `... By default this is the BIOS UUID`

pdellaert reviewed (changes requested):
Bunch of changes required. I'm also missing test cases for all the modules you have changed (because it would have identified a few issues, i think).

pdellaert commented:
Also think that this could be achieved with a boolean instead of a choice attribute, `is_instance_uuid` for those modules where a separate attribute makes sense (there's a bunch that already have `vm_id_type`, which should be reused)

cc @Akasurde

pdellaert commented:
@naphta, doing it transparent seems a bit dangerous. It's better, in my opinion, to give the choice. I do like the boolean approach more. I'd suggest `use_instance_uuid`, as it is more clear that it is a boolean and the purpose. Default false, of course.

Thanks!

naphta committed:
Update to boolean following review

naphta commented:
Had a quick run at converting it to use a boolean, I'll look at fleshing out the tests in the coming days. I'm not especially familar with the project but are there any tests beyond the integration ones I need to concern myself with?

naphta committed:
Fix docuementation errors raised

naphta committed:
Fix indentation error

naphta committed:
Fix error in integration test

naphta committed:
Fix documentation to reflect update

naphta commented:
I couldn't see any other integration tests which refer to the UUID other than the one I already included.

naphta committed:
Fix bug with using instance uuid’s

The above if statment was incorrectly including both instance and bios uuid’s, I’ve added a negate clause so each one will be hit correctly.

naphta commented:
ready_for_review

Akasurde committed:
VMware: Add tag facts to vmware_guest_facts (#47495)

Fixes: #46460

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

dagwieers commented:
@naphta Please answer all unresolved comments (and please also reply to each code-comment so it is clear that this has been taken into account/dismissed). Also please resolve the current conflict. Thanks!

naphta commented:
@dagwieers I believe all the comments were resolved, I'll look into the conflict and ensure all the comments have been replied to as well.

Cheers.

naphta commented:
@dagwieers I believe all the comments were resolved, I'll look into the conflict and ensure all the comments have been replied to as well.

Cheers.

dagwieers commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -2313,6 +2319,7 @@ def main():
         name=dict(type='str'),
         name_match=dict(type='str', choices=['first', 'last'], default='first'),
         uuid=dict(type='str'),
+        use_instance_uuid=dict(type='bool', default=False, required=False),
```
If the parameter is not required, it doesn't need this explicit. Please do this for all occurances.
```suggestion
        use_instance_uuid=dict(type='bool', default=False),
```

dagwieers replied:
Don't add `required=False`, as this is the default anyway. Do this for all occurrences.
```suggestion
        use_instance_uuid=dict(type='bool', default=False),
```

dagwieers commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -81,6 +81,12 @@
+    - Use the VMWare instance UUID rather than the BIOS UUID.
+    default: 'no'
+    type: bool
+    version_added: 2.7
```
This needs to become '2.8' now (as a string).
```suggestion
    version_added: '2.8'
```

dagwieers replied:
This should now be '2.8' (as a string). Please do this for all occurrences.
```suggestion
    version_added: '2.8'
```

dagwieers commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -81,6 +81,12 @@
+  use_instance_uuid:
+    description:
+    - Use the VMWare instance UUID rather than the BIOS UUID.
+    default: 'no'
```
Boolean values can now simply be yes or no. Please do this for all occurrences.
```suggestion
    default: no
```

dagwieers replied:
Please close comments that you have implemented. Outdated means the line has been changed, but doesn't guarantee that the comment was implemented. (In this case it's likely).

So please "resolve" implemented comments so it's clear what the status is of a review.

dagwieers commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -81,6 +81,12 @@
     - Please note that a supplied UUID will be ignored on virtual machine creation, as VMware creates the UUID internally.
+  use_instance_uuid:
+    description:
+    - Use the VMWare instance UUID rather than the BIOS UUID.
```
We usually start the description of boolean values with "Whether to use ...". Do this for all occurrences.
```suggestion
    - Whether to use the VMWare instance UUID rather than the BIOS UUID.
```

dagwieers reviewed (changes requested):
LGTM, except for some cosmetic/docs changes.

I would also like to see some of the integration tests using this option.

naphta committed:
Updates following review

dagwieers reviewed (approved):
LGTM, however I haven't tested this lacking a testing environment.

dagwieers commented:
@naphta Can you please rebase your branch, it seems CI has issues because the branch is too far behind. Go into your branch and run: `git pull --rebase origin devel`.

naphta commented:
Yep just in the process of doing that just now.

naphta commented:
Well that rebase didn't go so well. I'll revert my end and make the changes again.

dagwieers commented:
@naptha **Always** rebase using `git pull --rebase origin devel` so that your own commits are replayed after rebasing to devel. Otherwise your commit log cannot be properly merged with the devel branch and you end up in what you have.

naphta committed:
Fix issue with indentation of documentation

naphta commented:
> LGTM, except for some cosmetic/docs changes.
> 
> I would also like to see some of the integration tests using this option.

For what it's worth I did add a test in for this using the VMWare mock API included.

pdellaert reviewed (approved):
LGTM

Cool improvement :)

Akasurde commented:
Closing and re-opening for CI trigger.

Akasurde commented:
@naphta Thanks for the contribution. @pdellaert @dagwieers Thanks for the reviews.
